orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java

```diff
@@ -105,172 +105,120 @@ public long getTimeout() {
+      rollbackContexts.put(
+          region,
+          getRollbackContext(
+              stageData.targetHealthyRollbackPercentage, serverGroupToRollBack, candidateDetails));
+    }
 
-      if (rollbackDetails == null) {
-        // check for rollback candidates based on previous server groups
-        logger.info(
-            "Looking for rollback candidates in cluster {}, region {} based on previous server groups, "
-                + (onlyEnabled ? "excluding" : "including")
-                + " disabled ones",
-            moniker.get().getCluster(),
-            region);
-
-        rollbackDetails =
-            fetchRollbackDetails(
-                newestEnabledServerGroupInRegion,
-                onlyEnabled ? enabledServerGroupsInRegion : allServerGroupsInRegion);
-      }
+    return buildResult(imagesToRestore, rollbackTypes, rollbackContexts);
+  }
 
-      if (rollbackDetails != null) {
-        logger.info(
-            "Found rollback candidate in cluster {}, region {}: {}",
-            moniker.get().getCluster(),
-            region,
-            rollbackDetails.rollbackContext.get("restoreServerGroupName"));
+  private RollbackDetails findBestCandidate(
+      List<ServerGroup> allServerGroupsInRegion,
+      List<ServerGroup> enabledServerGroupsInRegion,
+      ServerGroup serverGroupToRollBack,
+      StageData stageData,
+      String cluster,
+      String region) {
 
-        Map<String, Object> rollbackContext = new HashMap<>(rollbackDetails.rollbackContext);
-        rollbackContext.put(
-            "targetHealthyRollbackPercentage",
-            determineTargetHealthyRollbackPercentage(
-                newestEnabledServerGroupInRegion.capacity,
-                stageData.targetHealthyRollbackPercentage));
+    List<ServerGroup> candidates =
+        shouldOnlyConsiderEnabledServerGroups(stageData.additionalRollbackContext)
+            ? enabledServerGroupsInRegion
+            : allServerGroupsInRegion;
 
-        rollbackTypes.put(region, rollbackDetails.rollbackType.toString());
-        rollbackContexts.put(region, rollbackContext);
+    ImageDetails imageDetails =
+        previousImageRollbackSupport.getImageDetailsFromEntityTags(
+            stageData.cloudProvider, stageData.credentials, region, serverGroupToRollBack.name);
 
-        ImmutableMap.Builder<Object, Object> imageToRestore =
-            ImmutableMap.builder()
-                .put("region", region)
-                .put("image", rollbackDetails.imageName)
-                .put("rollbackMethod", rollbackDetails.rollbackType.toString());
+    return getBestCandidate(cluster, region, serverGroupToRollBack, candidates, imageDetails);
+  }
 
-        if (rollbackDetails.buildNumber != null) {
-          imageToRestore.put("buildNumber", rollbackDetails.buildNumber);
-        }
+  private ServerGroup getServerGroupToRollBack(List<ServerGroup> enabledServerGroupsInRegion) {
```

Comment:
Consider annotating the parameter with `@Nonnull` since you're assuming that below.
```suggestion
  private ServerGroup getServerGroupToRollBack(@Nonnull List<ServerGroup> enabledServerGroupsInRegion) {
```